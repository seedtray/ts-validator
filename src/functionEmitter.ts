import {checkArgument} from './errors'
import {capitalize, unindent} from './stringUtils'
import {NamedType, Type} from './types'
import {typeToValidation} from './typeToValidation'
import {validationToJavascriptSource} from './validationToJavascriptSource'

/**
 * Emit validating functions.
 * for a type T, emit isT(value: unknown): value is T
 */
export class FunctionEmitter {

    typeToFunctionSource(t: NamedType, exportFunction: boolean = false): string {
        const expression = validationToJavascriptSource(typeToValidation(t.target), 'value')
        const functionName = this.getFunctionName(t.name)
        const exportPrefix = exportFunction ? 'export ' : ''

        return unindent(`
        ${exportPrefix}function ${functionName}(value: unknown): value is ${t.name} {
            return ${expression}
        }
        `)
    }

    private getFunctionName(typeName: string): string {
        const capitalizedTypeName = capitalize(typeName)

        return `is${capitalizedTypeName}`
    }
}
