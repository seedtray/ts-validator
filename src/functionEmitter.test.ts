import {FunctionEmitter} from './functionEmitter'
import {
    LiteralStringType,
    NamedType,
    nullableType,
    nullType,
    ObjectType,
    stringType,
    UnionType,
} from './types'

const fnEmitter = new FunctionEmitter()

test('emit validator for null', () => {
    const namedNull = NamedType.Of('Null', 'unknown', false, nullType)
    expect(fnEmitter.typeToFunctionSource(namedNull, false)).toMatchSnapshot()
    expect(fnEmitter.typeToFunctionSource(namedNull, true)).toMatchSnapshot()
})

test('emit validator for reasonably complicated object type', () => {
    const toValidate = NamedType.Of(
        'Lead',
        'unknown',
        false,
        ObjectType.Of({
            name: stringType,
            address: nullableType(stringType),
            kind: UnionType.Of([
                LiteralStringType.Of('business'),
                LiteralStringType.Of('person'),
            ]),
        }),
    )
    expect(fnEmitter.typeToFunctionSource(toValidate)).toMatchSnapshot()
})
