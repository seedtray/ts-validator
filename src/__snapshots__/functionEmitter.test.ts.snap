// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emit validator for null 1`] = `
"function isNull(value: unknown): value is Null {
    return value === null
}"
`;

exports[`emit validator for null 2`] = `
"export function isNull(value: unknown): value is Null {
    return value === null
}"
`;

exports[`emit validator for reasonably complicated object type 1`] = `
"function isLead(value: unknown): value is Lead {
    return ((Object(value) === value) && ((typeof value.name === 'string'
&& (value.address === null
|| typeof value.address === 'string')
&& (value.kind === \\"business\\"
|| value.kind === \\"person\\"))))
}"
`;
